File: anvil/src/test/java/trikita/anvil/MountTest.java
package trikita.anvil;

import android.view.View;

import org.junit.Test;
import static trikita.anvil.BaseDSL.*;
import org.mockito.Mockito;

import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import static junit.framework.Assert.*;

public class MountTest extends Utils {

    Anvil.Renderable testLayout = new Anvil.Renderable() {
        public void view() {
            v(MockView.class, new Anvil.Renderable() {
                public void view() {
                    attr("text", "bar");
                }
            });
        }
    };

    @Test
    public void testMountReturnsMountPoint() {
        assertEquals(container, Anvil.mount(container, empty));
    }

    @Test
    public void testMountRendersViews() {
        Anvil.mount(container, testLayout);
        assertEquals(1, container.getChildCount());
        assertTrue(container.getChildAt(0) instanceof MockView);
        MockView v = (MockView) container.getChildAt(0);
        assertEquals("bar", v.getText());
    }

    @Test
    public void testUnmountRemovesViews() {
        Anvil.mount(container, testLayout);
        assertEquals(1, container.getChildCount());
        Anvil.unmount(container);
        assertEquals(0, container.getChildCount());
    }

    @Test
    public void testMountReplacesViews() {
        Anvil.mount(container, testLayout);
        assertEquals(1, container.getChildCount());
        Anvil.mount(container, empty);
        assertEquals(0, container.getChildCount());
        Anvil.mount(container, testLayout);
        assertEquals(1, container.getChildCount());
    }

    @Test
    public void testMountInfoView() {
        MockView v = Anvil.mount(new MockView(getContext()), new Anvil.Renderable() {
            public void view() {
                attr("id", 100);
                attr("text", "bar");
                attr("tag", "foo");
            }
        });
        assertEquals(100, v.getId());
        assertEquals("foo", v.getTag());
        assertEquals("bar", v.getText());
    }

    @Test
    public void testMountGC() throws ExecutionException {
        Anvil.Renderable layout = Mockito.spy(testLayout);
        Anvil.mount(container, layout);
        Mockito.verify(layout, Mockito.times(1)).view();
        assertEquals(1, container.getChildCount());
        // Once the container is garbage collection all other views should be removed, too
        WeakReference<View> ref = new WeakReference<>(container.getChildAt(0));
        container = null;
        System.gc();
        assertEquals(null, ref.get());
        // Ensure that the associated renderable is no longer called
        renderView();
        Mockito.verify(layout, Mockito.times(1)).view();
    }

    private void renderView() throws ExecutionException {
        Future<?> future = Anvil.render();
        waitTaskToFinish(future);
    }

    private void waitTaskToFinish(Future<?> future) throws ExecutionException {
        try {
            future.get();
        } catch (ExecutionException executionException) {
            throw executionException;
        } catch (Exception e) {
        }
    }
}

File: anvil/src/test/kotlin/trikita/anvil/MountTest.kt
package trikita.anvil

import org.mockito.Mockito
import trikita.anvil.Anvil.Renderable
import java.lang.ref.WeakReference
import kotlin.test.*

class MountTest : Utils() {
    var testLayout = TestRenderable()

    @Test
    fun testMountReturnsMountPoint() {
        assertEquals(container, Anvil.mount(container, empty))
    }

    @Test
    fun testMountRendersViews() {
        Anvil.mount(container, testLayout)
        assertEquals(1, container!!.childCount)
        assertTrue(container!!.getChildAt(0) is MockView)
        val v = container!!.getChildAt(0) as MockView
        assertEquals("bar", v.text)
    }

    @Test
    fun testUnmountRemovesViews() {
        Anvil.mount(container, testLayout)
        assertEquals(1, container!!.childCount)
        Anvil.unmount(container)
        assertEquals(0, container!!.childCount)
    }

    @Test
    fun testMountReplacesViews() {
        Anvil.mount(container, testLayout)
        assertEquals(1, container!!.childCount)
        Anvil.mount(container, empty)
        assertEquals(0, container!!.childCount)
        Anvil.mount(container, testLayout)
        assertEquals(1, container!!.childCount)
    }

    @Test
    fun testMountInfoView() {
        val v = Anvil.mount(MockView(context)) {
            attr("id", 100)
            attr("text", "bar")
            attr("tag", "foo")
        }
        assertEquals(100, v.id)
        assertEquals("foo", v.tag)
        assertEquals("bar", v.text)
    }

    @Test
    fun testMountGC() {
        val layout = Mockito.spy(testLayout)
        Anvil.mount(container, layout)
        Mockito.verify(layout, Mockito.times(1)).view()
        assertEquals(1, container!!.childCount)
        // Once the container is garbage collection all other views should be removed, too
        val ref = WeakReference(container!!.getChildAt(0))
        container = null
        System.gc()
        assertEquals(null, ref.get())
        // Ensure that the associated renderable is no longer called
        Anvil.render()
        Mockito.verify(layout, Mockito.times(1)).view()
    }
}

open class TestRenderable : Renderable {
    override fun view() {
        v<Utils.MockView, ViewScope>(ViewScope) { attr("text", "bar") }
    }
}
